### DISCUSSION TITLE
domoticz/out subscriptions cause a lot of broadcast traffic

### DISCUSSION THREAD

**Discussion by aerofeev2k** [arendst/Tasmota#15677] [Q&A] [answered]
Hi,

I've tried Tasmota with Domoticz today for the first time, and couldn't help but notice that Tasmota subscribes to the entire domoticz/out topic. And that carries _a lot of_ traffic.

Every single measurement received, every single command is sent via domoticz/out, domoticz/out/$idx or domoticz/out/$name (depending on Domoticz configuration), and each one of them is getting picked up by each Tasmota instance because Tasmota subscribes to domoticz/out/#.

Granted, in the past Domoticz was pretty dumb to use domoticz/out unless there was a floor plan, so it was completely ignoring MQTT philosophy of using targeted topics. But now it grew up and can actually be configured to use proper per-device topics like domoticz/out/$idx or domoticz/out/$name, but... now Tasmota dropped the ball. It used to allow specifying in and out topics for Domoticz in the UI, so I could potentially make each ESP subscribe to the targeted domoticz/out/$name topic, but now these strings are hard-coded as domoticz/out/#.

Does anybody else see a problem with MQTT rebroadcasting every Domoticz sneeze to every Tasmota ESP device in the WiFi because Tasmota subscribes to domoticz/out/#?

Am I missing something big here?

Thanks!

**Comment by arendst** [comment]
As you noticed too domoticz was flooding MQTT with every message it wanted to send. Tasmota merely needed to read all messages and filter out the correct one based on the idx field.

MQTT doesn't rebroadcast messages. It broadcasts messages so other than that while domoticz sends a lot of messages it doesn't breakdown.

As far as I know tasmota has no issues picking up the correct message so I see no need for refining the topic in addition to domoticz tuning too.

**Reply by aerofeev2k** [reply]
> MQTT doesn't rebroadcast messages. It broadcasts messages so other than that while domoticz sends a lot of messages it doesn't breakdown.

That's not true anymore. Domoticz can be configured to use topics like domoticz/out/$idx where messages are sent to targeted devices via properly named topics. And I would be able to take advantage of this if Tasmota allowed me to configure the out topic per device. This functionality was there, but got removed.

Right now Tasmota subscribes to a wildcard topic, i.e. to every single message that Domoticz might send to other devices (and not necessarily to other Tasmotas). This is a lot of traffic. And filtering traffic on the end nodes goes against MQTT philosophy as well because end nodes are usually underpowered and are behind potentially unreliable network. Nodes should subscribe to a subset of messages they're interested in rather than listen to everybody else's chatter and then filter out the 1% of messages that are relevant.

**Comment by aerofeev2k** [comment, marked as answer]
I'd like to add a couple of points to this discussion...

One is that ESPurna actually allows user to do what I'm talking about, i.e. enter custom domoticz/out/$idx topic in ESP configuration thereby subscribing just to the messages intended for the device.

Second is that there is a workaround for Domoticz+Tasmota as well. Instead of using the "classic" MQTT Client Gateway with LAN interface in Domoticz, which communicates via domoticz/out/... and domoticz/in topics, one can use newer MQTT Auto Discovery Client Gateway with LAN interface and then enable MQTT Discovery in Tasmota via SetOption19. In this case no additional configuration is needed on the Tasmota side (and Tasmota's Domoticz section should in fact NOT be enabled), and Domoticz will auto-discover Tasmota's topics and will use them to send targeted communications.

**Reply by aerofeev2k** [reply]
And this is how the old (but improved) Domoticz "MQTT Client Gateway with LAN interface" sends out messages these days:

```
domoticz/out/183 {
	"Battery" : 255,
	"LastUpdate" : "2022-05-25 11:00:46",
	"RSSI" : 12,
	"description" : "",
	"dtype" : "Light/Switch",
	"hwid" : "5",
	"id" : "00014107",
	"idx" : 183,
	"name" : "BookShelfSwitch",
	"nvalue" : 0,
	"stype" : "Switch",
	"svalue1" : "0",
	"switchType" : "On/Off",
	"unit" : 1
}
```

Note that it can now be configured to use a device-specific topic domoticz/out/183.

**Reply by arendst** [reply]
So in essence tasmota should change from sending MQTT data from ``domoticz/in``  to ``domoticz/in/<idx>`` where idx is currently send with the command message.

Same for ``domoticz/out`` detection. It should look for ``domoticz/out/<idx>`` where the idx is defined in the domoticz configuration GUI page or command ``DzIdx``.

Pls verify. I will then do some tests and see how to release this with as few as possible user interaction (considering backward compatibility).

**Reply by arendst** [reply]
BTW with what version of Domoticz is this supported? Does the current released version support this?

**Reply by arendst** [reply]
Ah I see. The current version does support it: 
![image](https://user-images.githubusercontent.com/11044339/170732083-16c997ea-57d6-4faa-a74d-0b52916341a7.png)

**Reply by arendst** [reply]
I don't see if it works with ``domoticz/in/<idx>`` too ????

**Reply by arendst** [reply]
Anyway. It doesn't change domoticz flooding the MQTT server. It merely adds the idx to the previous topic so other than that Tasmota can filter out the ``domoticz/out/<idx>`` easier so I see no real advantage (as Tasmota was always able to filter out the correct idx without hickups)

**Reply by aerofeev2k** [reply]
> I don't see if it works with `domoticz/in/<idx>` too ????

Correct, Domoticz still expects to be talked to via domoticz/in only.

**Reply by aerofeev2k** [reply]
> Anyway. It doesn't change domoticz flooding the MQTT server. It merely adds the idx to the previous topic so other than that Tasmota can filter out the `domoticz/out/<idx>` easier so I see no real advantage (as Tasmota was always able to filter out the correct idx without hickups)

This is also correct. Domoticz will still send messages for all device changes to the MQTT server, but that's not a problem. MQTT is the broker that's supposed to scale and that is likely sitting on the same machine or close to Domoticz, so it can handle the flood.

The problem is that if a number of Tasmota devices all subscribe to domoticz/out/# topic, then each of the Tasmota devices will be receiving traffic that's intended for somebody else. So that flood will multiply N times by being forked to each of the devices. If on the other hand Tasmota subscribes to a narrowed topic like domocitz/out/$idx, then MQTT server will forward only relevant messages to each of the Tasmota devices.

**Reply by aerofeev2k** [reply]
> One is that ESPurna actually allows user to do what I'm talking about, i.e. enter custom domoticz/out/$idx topic in ESP configuration thereby subscribing just to the messages intended for the device.

I realized later that while ESPurna's way actually works nicely out of the box for single relay devices if one manually types a topic like the above, it fails if there are multiple relays, because multiple subscriptions would be required.

So some code changes would be needed anyway. For Tasmota in my network I just did this:

```
diff --git a/tasmota/xdrv_07_domoticz.ino b/tasmota/xdrv_07_domoticz.ino
index b469ef1d5..105a12d52 100644
--- a/tasmota/xdrv_07_domoticz.ino
+++ b/tasmota/xdrv_07_domoticz.ino
@@ -175,18 +175,19 @@ void DomoticzMqttUpdate(void) {
 }
 
 void DomoticzMqttSubscribe(void) {
-  uint8_t maxdev = (TasmotaGlobal.devices_present > MAX_DOMOTICZ_IDX) ? MAX_DOMOTICZ_IDX : TasmotaGlobal.devices_present;
-  for (uint32_t i = 0; i < maxdev; i++) {
-    if (Settings->domoticz_relay_idx[i]) {
-      domoticz_subscribe = true;
-    }
-  }
-
-  if (domoticz_subscribe) {
-    char stopic[TOPSZ];
-    snprintf_P(stopic, sizeof(stopic), PSTR(DOMOTICZ_OUT_TOPIC "/#")); // domoticz topic
-    MqttSubscribe(stopic);
-  }
+       /* Since my Domoticz is configured to publish in domoticz/out/idx,
+        * we want to subscribe to every switch's topic (and there can be multiple,
+        * if there are multiple switches in the device). */
+       uint8_t maxdev = (TasmotaGlobal.devices_present > MAX_DOMOTICZ_IDX) ? MAX_DOMOTICZ_IDX : TasmotaGlobal.devices_present;
+       for (uint32_t i = 0; i < maxdev; i++) {
+               int idx=Settings->domoticz_relay_idx[i];
+               if (idx) {
+                       char stopic[TOPSZ];
+                       snprintf_P(stopic, sizeof(stopic), PSTR(DOMOTICZ_OUT_TOPIC "/%d"), idx);
+                       MqttSubscribe(stopic);
+                       domoticz_subscribe = true;
+               }
+       }
 }
 
 bool DomoticzMqttData(void) {
```

But having something like this in the official firmware would require user's confirmation that Domoticz is indeed configured to use the domoticz/out/$idx topics to communicate.

**Reply by arendst** [reply]
I add a compromise speeding up detection of correct idx:

```
  // Quick check if this is mine using topic domoticz/out/{$idx}
  if (strlen(XdrvMailbox.topic) > strlen(DOMOTICZ_OUT_TOPIC)) {
    char* topic_index = &XdrvMailbox.topic[strlen(DOMOTICZ_OUT_TOPIC) +1];
    uint32_t idx = atoi(topic_index);
    uint8_t maxdev = (TasmotaGlobal.devices_present > MAX_DOMOTICZ_IDX) ? MAX_DOMOTICZ_IDX : TasmotaGlobal.devices_present;
    bool not_found = true;
    for (uint32_t i = 0; i < maxdev; i++) {
      if (idx == Settings->domoticz_relay_idx[i]) {
        not_found = false;
        break;
      }
    }
    if (not_found) {
      return true;  // Idx not mine
    }
  }
```

Added to latest dev release just now.

**Comment by joba-1** [comment]
I think the main point for a more specific subscription is, mqtt would not need to send the messages to all tasmota devices and flood the wlan
